src/Symfony/Component/DependencyInjection/LazyProxy/ProxyHelper.php

```diff
@@ -33,22 +33,31 @@ public static function getTypeHint(\ReflectionFunctionAbstract $r, \ReflectionPa
         }
 
         $types = [];
+        $glue = '|';
+        if ($type instanceof \ReflectionUnionType) {
+            $reflectionTypes = $type->getTypes();
+        } elseif ($type instanceof \ReflectionIntersectionType) {
+            $reflectionTypes = $type->getTypes();
+            $glue = '&';
+        } elseif ($type instanceof \ReflectionNamedType) {
+            $reflectionTypes = [$type];
+        } else {
+            return null;
+        }
 
-        foreach ($type instanceof \ReflectionUnionType ? $type->getTypes() : [$type] as $type) {
-            $name = $type instanceof \ReflectionNamedType ? $type->getName() : (string) $type;
```

Comment:
on 7.1, don't we need the code path with a cast to string?

Reply:
The string cast was necessary for PHP 7.0 and earlier.

The `if` block above makes sure we have an array of [`ReflectionNamedType`](https://www.php.net/ReflectionNamedType):
* It is the only core implementation of `ReflectionType` on PHP 7.1 - 7.4.
* The `getTypes()` method on `ReflectionUnionType` and `ReflectionIntersectionType` will (as of today) return an array of `ReflectionNamedType`.
* If we encounter another implementation (which a future PHP release might introduce), this method will give up and exit (this is the currently unreachable `else return null;` branch).

It is possible to create userland implementations of `ReflectionType` though, but at this point we should only be dealing with reflection objects coming from PHP directly.